<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Overview of the tidySummarizedExperiment package • tidySummarizedExperiment</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the tidySummarizedExperiment package">
<meta property="og:description" content="tidySummarizedExperiment">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidySummarizedExperiment</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.99.3</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fas fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/introduction.html">Overview of the tidySummarizedExperiment package</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right"></ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="introduction_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Overview of the tidySummarizedExperiment package</h1>
                        <h4 class="author">Stefano Mangiola</h4>
            
            <h4 class="date">2021-02-10</h4>
      
      
      <div class="hidden name"><code>introduction.Rmd</code></div>

    </div>

    
    
<!-- badges: start -->
<p><a href="https://www.tidyverse.org/lifecycle/#maturing"><img src="https://img.shields.io/badge/lifecycle-maturing-blue.svg" alt="Lifecycle:maturing"></a> <!-- badges: end --></p>
<p><strong>Brings SummarizedExperiment to the tidyverse!</strong></p>
<p>website: <a href="https://stemangiola.github.io/tidySummarizedExperiment/">stemangiola.github.io/tidySummarizedExperiment/</a></p>
<p>Please also have a look at</p>
<ul>
<li>
<a href="https://stemangiola.github.io/tidySCE/">tidySCE</a> for tidy manipulation of Seurat objects</li>
<li>
<a href="https://stemangiola.github.io/tidyseurat/">tidyseurat</a> for tidy manipulation of Seurat objects</li>
<li>
<a href="https://stemangiola.github.io/tidybulk/">tidybulk</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/nanny">nanny</a> for tidy high-level data analysis and manipulation</li>
<li>
<a href="https://github.com/stemangiola/tidygate">tidygate</a> for adding custom gate information to your tibble</li>
<li>
<a href="https://stemangiola.github.io/tidyHeatmap/">tidyHeatmap</a> for heatmaps produced with tidy principles</li>
</ul>
<div id="introduction" class="section level1">
<h1 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h1>
<p>tidySummarizedExperiment provides a bridge between Bioconductor <a href="https://bioconductor.org/packages/release/bioc/html/SummarizedExperiment.html">SummarizedExperiment</a> <span class="citation">(Morgan et al. 2020)</span> and the tidyverse <span class="citation">(Wickham et al. 2019)</span>. It creates an invisible layer that enables viewing the Bioconductor <em>SummarizedExperiment</em> object as a tidyverse tibble, and provides SummarizedExperiment-compatible <em>dplyr</em>, <em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions. This allows users to get the best of both Bioconductor and tidyverse worlds.</p>
<div id="functionsutilities-available" class="section level2">
<h2 class="hasAnchor">
<a href="#functionsutilities-available" class="anchor"></a>Functions/utilities available</h2>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>SummarizedExperiment-compatible Functions</th>
<th>Description</th>
</tr></thead>
<tbody><tr class="odd">
<td><code>all</code></td>
<td>After all <code>tidySummarizedExperiment</code> is a SummarizedExperiment object, just better</td>
</tr></tbody>
</table>
<table class="table">
<thead><tr class="header">
<th>tidyverse Packages</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>dplyr</code></td>
<td>Almost all <code>dplyr</code> APIs like for any tibble</td>
</tr>
<tr class="even">
<td><code>tidyr</code></td>
<td>Almost all <code>tidyr</code> APIs like for any tibble</td>
</tr>
<tr class="odd">
<td><code>ggplot2</code></td>
<td>
<code>ggplot</code> like for any tibble</td>
</tr>
<tr class="even">
<td><code>plotly</code></td>
<td>
<code>plot_ly</code> like for any tibble</td>
</tr>
</tbody>
</table>
<table class="table">
<colgroup>
<col width="48%">
<col width="52%">
</colgroup>
<thead><tr class="header">
<th>Utilities</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>tidy</code></td>
<td>Add <code>tidySummarizedExperiment</code> invisible layer over a SummarizedExperiment object</td>
</tr>
<tr class="even">
<td><code>as_tibble</code></td>
<td>Convert cell-wise information to a <code>tbl_df</code>
</td>
</tr>
</tbody>
</table>
</div>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span>
      <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span>
  <span class="op">}</span>

<span class="fu">BiocManager</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/BiocManager/man/install.html">install</a></span><span class="op">(</span><span class="st">"tidySummarizedExperiment"</span><span class="op">)</span></code></pre></div>
<p>From Github (development)</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">devtools</span><span class="fu">::</span><span class="fu"><a href="https://devtools.r-lib.org//reference/remote-reexports.html">install_github</a></span><span class="op">(</span><span class="st">"stemangiola/tidySummarizedExperiment"</span><span class="op">)</span></code></pre></div>
<p>Load libraries used in the examples.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org">ggplot2</a></span><span class="op">)</span>
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span><span class="op">(</span><span class="va">tidySummarizedExperiment</span><span class="op">)</span></code></pre></div>
</div>
</div>
<div id="create-tidysummarizedexperiment-the-best-of-both-worlds" class="section level1">
<h1 class="hasAnchor">
<a href="#create-tidysummarizedexperiment-the-best-of-both-worlds" class="anchor"></a>Create <code>tidySummarizedExperiment</code>, the best of both worlds!</h1>
<p>This is a SummarizedExperiment object but it is evaluated as a tibble. So it is fully compatible both with SummarizedExperiment and tidyverse APIs.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">&lt;-</span> <span class="fu">tidySummarizedExperiment</span><span class="fu">::</span><span class="va"><a href="../reference/pasilla.html">pasilla</a></span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/tidy.html">tidy</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<p><strong>It looks like a tibble</strong></p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span></code></pre></div>
<pre><code>## # A tibble abstraction: 102,193 x 5
##    sample condition type       transcript  counts
##    &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;
##  1 untrt1 untreated single_end FBgn0000003      0
##  2 untrt1 untreated single_end FBgn0000008     92
##  3 untrt1 untreated single_end FBgn0000014      5
##  4 untrt1 untreated single_end FBgn0000015      0
##  5 untrt1 untreated single_end FBgn0000017   4664
##  6 untrt1 untreated single_end FBgn0000018    583
##  7 untrt1 untreated single_end FBgn0000022      0
##  8 untrt1 untreated single_end FBgn0000024     10
##  9 untrt1 untreated single_end FBgn0000028      0
## 10 untrt1 untreated single_end FBgn0000032   1446
## # … with 40 more rows</code></pre>
<p><strong>But it is a SummarizedExperiment object after all</strong></p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu">Assays</span><span class="op">(</span><span class="va">pasilla_tidy</span><span class="op">)</span></code></pre></div>
<pre><code>## An object of class "SimpleAssays"
## Slot "data":
## List of length 1</code></pre>
</div>
<div id="tidyverse-commands" class="section level1">
<h1 class="hasAnchor">
<a href="#tidyverse-commands" class="anchor"></a>Tidyverse commands</h1>
<p>We can use tidyverse commands to explore the tidy SummarizedExperiment object.</p>
<p>We can use <code>slice</code> to choose rows by position, for example to choose the first row.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble abstraction: 1 x 5
##   sample condition type       transcript  counts
##   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;
## 1 untrt1 untreated single_end FBgn0000003      0</code></pre>
<p>We can use <code>filter</code> to choose rows by criteria.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">filter</a></span><span class="op">(</span><span class="va">condition</span> <span class="op">==</span> <span class="st">"untreated"</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble abstraction: 58,396 x 5
##    sample condition type       transcript  counts
##    &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;
##  1 untrt1 untreated single_end FBgn0000003      0
##  2 untrt1 untreated single_end FBgn0000008     92
##  3 untrt1 untreated single_end FBgn0000014      5
##  4 untrt1 untreated single_end FBgn0000015      0
##  5 untrt1 untreated single_end FBgn0000017   4664
##  6 untrt1 untreated single_end FBgn0000018    583
##  7 untrt1 untreated single_end FBgn0000022      0
##  8 untrt1 untreated single_end FBgn0000024     10
##  9 untrt1 untreated single_end FBgn0000028      0
## 10 untrt1 untreated single_end FBgn0000032   1446
## # … with 40 more rows</code></pre>
<p>We can use <code>select</code> to choose columns.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">select</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 102,193 x 1
##    sample
##    &lt;chr&gt; 
##  1 untrt1
##  2 untrt1
##  3 untrt1
##  4 untrt1
##  5 untrt1
##  6 untrt1
##  7 untrt1
##  8 untrt1
##  9 untrt1
## 10 untrt1
## # … with 102,183 more rows</code></pre>
<p>We can use <code>count</code> to count how many rows we have for each sample.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/count.html">count</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 7 x 2
##   sample     n
## * &lt;chr&gt;  &lt;int&gt;
## 1 trt1   14599
## 2 trt2   14599
## 3 trt3   14599
## 4 untrt1 14599
## 5 untrt2 14599
## 6 untrt3 14599
## 7 untrt4 14599</code></pre>
<p>We can use <code>distinct</code> to see what distinct sample information we have.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">distinct</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">condition</span>, <span class="va">type</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 7 x 3
##   sample condition type      
##   &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;     
## 1 untrt1 untreated single_end
## 2 untrt2 untreated single_end
## 3 untrt3 untreated paired_end
## 4 untrt4 untreated paired_end
## 5 trt1   treated   single_end
## 6 trt2   treated   paired_end
## 7 trt3   treated   paired_end</code></pre>
<p>We could use <code>rename</code> to rename a column. For example, to modify the type column name.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">rename</a></span><span class="op">(</span>sequencing<span class="op">=</span><span class="va">type</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble abstraction: 102,193 x 5
##    sample condition sequencing transcript  counts
##    &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;
##  1 untrt1 untreated single_end FBgn0000003      0
##  2 untrt1 untreated single_end FBgn0000008     92
##  3 untrt1 untreated single_end FBgn0000014      5
##  4 untrt1 untreated single_end FBgn0000015      0
##  5 untrt1 untreated single_end FBgn0000017   4664
##  6 untrt1 untreated single_end FBgn0000018    583
##  7 untrt1 untreated single_end FBgn0000022      0
##  8 untrt1 untreated single_end FBgn0000024     10
##  9 untrt1 untreated single_end FBgn0000028      0
## 10 untrt1 untreated single_end FBgn0000032   1446
## # … with 40 more rows</code></pre>
<p>We could use <code>mutate</code> to create a column. For example, we could create a new type column that contains single and paired instead of single_end and paired_end.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">mutate</a></span><span class="op">(</span>type<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html">gsub</a></span><span class="op">(</span><span class="st">"_end"</span>, <span class="st">""</span>, <span class="va">type</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble abstraction: 102,193 x 5
##    sample condition type   transcript  counts
##    &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;  &lt;chr&gt;        &lt;int&gt;
##  1 untrt1 untreated single FBgn0000003      0
##  2 untrt1 untreated single FBgn0000008     92
##  3 untrt1 untreated single FBgn0000014      5
##  4 untrt1 untreated single FBgn0000015      0
##  5 untrt1 untreated single FBgn0000017   4664
##  6 untrt1 untreated single FBgn0000018    583
##  7 untrt1 untreated single FBgn0000022      0
##  8 untrt1 untreated single FBgn0000024     10
##  9 untrt1 untreated single FBgn0000028      0
## 10 untrt1 untreated single FBgn0000032   1446
## # … with 40 more rows</code></pre>
<p>We could use <code>unite</code> to combine multiple columns into a single column.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/unnest.html">unite</a></span><span class="op">(</span><span class="st">"group"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html">c</a></span><span class="op">(</span><span class="va">condition</span>, <span class="va">type</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble abstraction: 102,193 x 4
##    sample group                transcript  counts
##    &lt;chr&gt;  &lt;chr&gt;                &lt;chr&gt;        &lt;int&gt;
##  1 untrt1 untreated_single_end FBgn0000003      0
##  2 untrt1 untreated_single_end FBgn0000008     92
##  3 untrt1 untreated_single_end FBgn0000014      5
##  4 untrt1 untreated_single_end FBgn0000015      0
##  5 untrt1 untreated_single_end FBgn0000017   4664
##  6 untrt1 untreated_single_end FBgn0000018    583
##  7 untrt1 untreated_single_end FBgn0000022      0
##  8 untrt1 untreated_single_end FBgn0000024     10
##  9 untrt1 untreated_single_end FBgn0000028      0
## 10 untrt1 untreated_single_end FBgn0000032   1446
## # … with 40 more rows</code></pre>
<p>We can also combine commands with the tidyverse pipe <code><a href="../reference/%&gt;%.html">%&gt;%</a></code>.</p>
<p>For example, we could combine <code>group_by</code> and <code>summarise</code> to get the total counts for each sample.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">group_by</a></span><span class="op">(</span><span class="va">sample</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">summarise</a></span><span class="op">(</span>total_counts<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/sum.html">sum</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 7 x 2
##   sample total_counts
## * &lt;chr&gt;         &lt;int&gt;
## 1 trt1       18670279
## 2 trt2        9571826
## 3 trt3       10343856
## 4 untrt1     13972512
## 5 untrt2     21911438
## 6 untrt3      8358426
## 7 untrt4      9841335</code></pre>
<p>We could combine <code>group_by</code>, <code>mutate</code> and <code>filter</code> to get the transcripts with mean count &gt; 0.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">group_by</a></span><span class="op">(</span><span class="va">transcript</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">mutate</a></span><span class="op">(</span>mean_count<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/mean.html">mean</a></span><span class="op">(</span><span class="va">counts</span><span class="op">)</span><span class="op">)</span> <span class="op">%&gt;%</span>
    <span class="fu"><a href="../reference/bind_rows.html">filter</a></span><span class="op">(</span><span class="va">mean_count</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span></code></pre></div>
<pre><code>## # A tibble: 86,513 x 6
## # Groups:   transcript [12,359]
##    sample condition type       transcript  counts mean_count
##    &lt;chr&gt;  &lt;chr&gt;     &lt;chr&gt;      &lt;chr&gt;        &lt;int&gt;      &lt;dbl&gt;
##  1 untrt1 untreated single_end FBgn0000003      0      0.143
##  2 untrt1 untreated single_end FBgn0000008     92     99.6  
##  3 untrt1 untreated single_end FBgn0000014      5      1.43 
##  4 untrt1 untreated single_end FBgn0000015      0      0.857
##  5 untrt1 untreated single_end FBgn0000017   4664   4672.   
##  6 untrt1 untreated single_end FBgn0000018    583    461.   
##  7 untrt1 untreated single_end FBgn0000022      0      0.143
##  8 untrt1 untreated single_end FBgn0000024     10      7    
##  9 untrt1 untreated single_end FBgn0000028      0      0.429
## 10 untrt1 untreated single_end FBgn0000032   1446   1085.   
## # … with 86,503 more rows</code></pre>
</div>
<div id="plotting" class="section level1">
<h1 class="hasAnchor">
<a href="#plotting" class="anchor"></a>Plotting</h1>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">my_theme</span> <span class="op">&lt;-</span>
    <span class="fu"><a href="https://rdrr.io/r/base/list.html">list</a></span><span class="op">(</span>
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_fill_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Set1"</span><span class="op">)</span>,
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html">scale_color_brewer</a></span><span class="op">(</span>palette<span class="op">=</span><span class="st">"Set1"</span><span class="op">)</span>,
        <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggtheme.html">theme_bw</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
            <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html">theme</a></span><span class="op">(</span>
                panel.border<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
                axis.line<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span><span class="op">)</span>,
                panel.grid.major<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.2</span><span class="op">)</span>,
                panel.grid.minor<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_line</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.1</span><span class="op">)</span>,
                text<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">12</span><span class="op">)</span>,
                legend.position<span class="op">=</span><span class="st">"bottom"</span>,
                aspect.ratio<span class="op">=</span><span class="fl">1</span>,
                strip.background<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_blank</a></span><span class="op">(</span><span class="op">)</span>,
                axis.title.x<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t<span class="op">=</span><span class="fl">10</span>, r<span class="op">=</span><span class="fl">10</span>, b<span class="op">=</span><span class="fl">10</span>, l<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>,
                axis.title.y<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">element_text</a></span><span class="op">(</span>margin<span class="op">=</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html">margin</a></span><span class="op">(</span>t<span class="op">=</span><span class="fl">10</span>, r<span class="op">=</span><span class="fl">10</span>, b<span class="op">=</span><span class="fl">10</span>, l<span class="op">=</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span>
            <span class="op">)</span>
    <span class="op">)</span></code></pre></div>
<p>We can treat <code>pasilla_tidy</code> as a normal tibble for plotting.</p>
<p>Here we plot the distribution of counts per sample.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pasilla_tidy</span> <span class="op">%&gt;%</span>
    <span class="fu">tidySummarizedExperiment</span><span class="fu">::</span><span class="fu"><a href="../reference/ggplot.html">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html">aes</a></span><span class="op">(</span><span class="va">counts</span> <span class="op">+</span> <span class="fl">1</span>, group<span class="op">=</span><span class="va">sample</span>, color<span class="op">=</span><span class="va">`type`</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_density.html">geom_density</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html">scale_x_log10</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span>
    <span class="va">my_theme</span></code></pre></div>
<p><img src="introduction_files/figure-html/plot1-1.png" width="700"></p>
</div>
<div id="session-info" class="section level1">
<h1 class="hasAnchor">
<a href="#session-info" class="anchor"></a>Session Info</h1>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></code></pre></div>
<pre><code>## R version 4.0.3 (2020-10-10)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 20.04.1 LTS
## 
## Matrix products: default
## BLAS/LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.8.so
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
##  [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=C             
##  [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       
## 
## attached base packages:
## [1] parallel  stats4    stats     graphics  grDevices utils     datasets 
## [8] methods   base     
## 
## other attached packages:
##  [1] tidySummarizedExperiment_0.99.3 SummarizedExperiment_1.20.0    
##  [3] Biobase_2.50.0                  GenomicRanges_1.42.0           
##  [5] GenomeInfoDb_1.26.2             IRanges_2.24.1                 
##  [7] S4Vectors_0.28.1                BiocGenerics_0.36.0            
##  [9] MatrixGenerics_1.2.1            matrixStats_0.58.0             
## [11] ggplot2_3.3.3                   knitr_1.31                     
## [13] BiocStyle_2.18.1               
## 
## loaded via a namespace (and not attached):
##  [1] lattice_0.20-41        tidyr_1.1.2            utf8_1.1.4            
##  [4] assertthat_0.2.1       rprojroot_2.0.2        digest_0.6.27         
##  [7] R6_2.5.0               evaluate_0.14          highr_0.8             
## [10] httr_1.4.2             pillar_1.4.7           zlibbioc_1.36.0       
## [13] rlang_0.4.10           lazyeval_0.2.2         data.table_1.13.6     
## [16] Matrix_1.3-2           rmarkdown_2.6          pkgdown_1.6.1         
## [19] labeling_0.4.2         textshaping_0.2.1      desc_1.2.0            
## [22] stringr_1.4.0          htmlwidgets_1.5.3      RCurl_1.98-1.2        
## [25] munsell_0.5.0          DelayedArray_0.16.1    compiler_4.0.3        
## [28] xfun_0.20              pkgconfig_2.0.3        systemfonts_1.0.0     
## [31] htmltools_0.5.1.1      tidyselect_1.1.0       tibble_3.0.6          
## [34] GenomeInfoDbData_1.2.4 bookdown_0.21          viridisLite_0.3.0     
## [37] fansi_0.4.2            crayon_1.4.1           dplyr_1.0.4           
## [40] withr_2.4.1            bitops_1.0-6           grid_4.0.3            
## [43] jsonlite_1.7.2         gtable_0.3.0           lifecycle_0.2.0       
## [46] magrittr_2.0.1         scales_1.1.1           cli_2.3.0             
## [49] stringi_1.5.3          cachem_1.0.3           farver_2.0.3          
## [52] XVector_0.30.0         fs_1.5.0               ellipsis_0.3.1        
## [55] ragg_0.4.1             generics_0.1.0         vctrs_0.3.6           
## [58] RColorBrewer_1.1-2     tools_4.0.3            glue_1.4.2            
## [61] purrr_0.3.4            fastmap_1.1.0          yaml_2.2.1            
## [64] colorspace_2.0-0       BiocManager_1.30.10    memoise_2.0.0         
## [67] plotly_4.9.3</code></pre>
</div>
<div id="references" class="section level1 unnumbered">
<h1 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h1>
<div id="refs" class="references">
<div id="ref-morgan2020summarized">
<p>Morgan, Martin, Valerie Obenchain, Jim Hester, and Hervé Pagès. 2020. <em>SummarizedExperiment: SummarizedExperiment Container</em>.</p>
</div>
<div id="ref-wickham2019welcome">
<p>Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019. “Welcome to the Tidyverse.” <em>Journal of Open Source Software</em> 4 (43): 1686.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Stefano Mangiola.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.6.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
